lib/wallet/txdb.js

```diff
@@ -79,6 +79,52 @@ class TXDB {
     this.wallet.emit(event, data, details);
   }
 
+  /**
+   * Emit events for each type of covenant.
+   * @private
+   * @param {TXRecord} wtx
+   * @param {Details} details
+   */
+
+  async emitCovenants(tx, details) {
+    const {outputs} = tx;
+    for (const output of outputs) {
+      const {covenant} = output;
+      const {type} = covenant;
+      const action = typesByVal[type].toLowerCase();
+      const channel = `${action} covenant`;
+      switch (type) {
+        case types.CLAIM:
+        case types.OPEN:
+        case types.BID:
+        case types.FINALIZE: {
+          const name = covenant.get(2).toString();
+          this.emit(channel, {name, output, tx}, details);
+          break;
+        }
+        case types.REVEAL:
```

Comment:
For these covenants, the name is not in the covenant data, it requires a db lookup to look up the name based on the hash. I don't think this will result in too much io

Reply:
To alleviate the added i/o, we could refactor [`this.connectNames`][1]. Currently, after processing incoming covenants, the method returns a boolean that indicates whether the name state was updated. We could instead have it return a summary object which accumulates all processed covenants.

We could then pass this summary object to `this.emitCovenants`.

[1]: https://github.com/handshake-org/hsd/blob/e1c0281dd9e9bbbd23f5cd832c82ebe3808e9b4f/lib/wallet/txdb.js#L1776

Reply:
See https://github.com/handshake-org/hsd/pull/195#discussion_r296451469